doc/development/powershell.md

```diff
@@ -0,0 +1,179 @@
+- Isolate calls to dependencies inside their own cmdlet files or within functions. This makes it easier to abstract these calls for testing.
+- Avoid global state where possible. Powershell has very liberal scoping rules, and it is easy to take dependencies on variables that are defined outside a code block, which can lead to hard to debug issues. Where possible, only treat script parameters and variables defined at the top of the file as read-only global variables.
+- Try not to rely on exporting your code as a powershell module. Favor [dot sourcing](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_scripts?view=powershell-7.2#script-scope-and-dot-sourcing) when possible. There are lots of downsides to working with module imports in local development settings that make them not worth it. This guideline some limitations, e.g. if you want to test class instance assertions via Pester, you cannot export the class types without using module imports.
+
+## Functionality
+
+There are many quirks to how powershell works that make trade offs between being a good command line shell vs. being a good scripting language. Here are some things to keep in mind when writing scripts:
+
+- Reserve powershell pipelines (`|` operator) for simple assignment operations only (e.g. light transformations of a data structure)
+  - If side effects are needed, favor `foreach` loops.
+  - If the value being returned is any sort of collection (e.g. list), either favor a `foreach` loop or explicitly declare a variable type like `[array]$arr = some-func | ? { $_ -contains "foo" }` . Due to how pipelines work, when the result value is singular, the inner value with the collection item type will be returned instead of a collection with length 1. _This goes the same for powershell functions._
+- Handle exit codes from calls to external binaries (e.g. `git`).
+  - Powershell by default will only stop execution on thrown exceptions. Calls to external binaries do not throw powershell native exceptions, so failures will not be caught unless explicitly handled.
+    - Check the `$LASTEXITCODE` variable for values greater than 0.
+    - Use the new `$PSNativeCommandUseErrorActionPreference` variable coming in 7.3 (https://docs.microsoft.com/en-us/powershell/scripting/whats-new/what-s-new-in-powershell-73?view=powershell-7.2#experimental-features).
+- Set global powershell variables to achieve consistent behavior:
+  - `$ErrorActionPreference` should be set to `Stop` at the top of any cmdlet files unless there is a specific reason to do something else.
+  - `Set-StrictMode -Version 4` should be set at the top of any cmdlet files unless there is a specific reason to do something else.
+- Support the common `-WhatIf` parameter. See [SupportsShouldProcess docs](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-shouldprocess?view=powershell-7.2#supportsshouldprocess).
```

Comment:
You will need to have CmdBinding set on the script parameters to get this common parameter, which is why I also mentioned it in an earlier comment.